sympy/sets/fancysets.py

```diff
@@ -634,17 +634,33 @@ def __new__(cls, sets, polar=False):
 
         # Rectangular Form
         if polar is False:
-            obj = ImageSet.__new__(cls, Lambda((x, y), x + I*y), sets)
+
+            if (all([_a.is_FiniteSet for _a in sets.args]) and
+               all([len(_a) == 1 for _a in sets.args])):
```

Comment:
You don't need the `[]` here, all() will work fine on a generator, and will make it not have to create a list. You can also combine these with `_a.is_FiniteSet and len(_a) == 1` (tho that may not necessarily help performance)